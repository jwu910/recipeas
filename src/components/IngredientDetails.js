import React, { Component } from 'react';
import {
  ListGroup, Button, FormControl, ControlLabel
} from 'react-bootstrap';
import { Link, Redirect } from 'react-router-dom';
import { searchNutrientInfo, roundToTwo } from '../utils/api';
import NutrientsList from './NutrientsList';
import PropTypes from 'prop-types';

//gets dbNumber from props.location.state.dbNumber in the form of:
// {
//   dbNumber: '1234567'
// }

class IngredientDetails extends Component {
  constructor(props) {
    super(props);

    this.state = {
      originalDetails: null,
      updatedDetails: null,
      servings: 1,
      redirect: false
    };

    this.renderNutrientDetails = this.renderNutrientDetails.bind(this);
    this.handleServingsChange = this.handleServingsChange.bind(this);
  }
  renderNutrientDetails() {
    return (
      <div>
        <h4>{this.state.updatedDetails.name}</h4>
        <ControlLabel>Servings</ControlLabel>
        <FormControl
          type="number"
          placeholder="1"
          value={this.state.servings}
          onChange={this.handleServingsChange}
        />
        <p><strong>Serving Size:</strong> {this.state.updatedDetails.measure}</p>

        <ListGroup>
          <h4>Macronutrients</h4>
          <NutrientsList
            nutrients={this.state.updatedDetails.macronutrients}
          />
          <h4>Micronutrients</h4>
          <NutrientsList
            nutrients={this.state.updatedDetails.micronutrients}
          />
        </ListGroup>

        <Link to="/modify-ingredients">
          <Button
            bsSize="large"
            bsStyle="primary"
            block
            style={{ marginBottom: '10px' }}
            onClick={() => {
              this.props.addTempIngredient(this.state.updatedDetails);
              this.props.clearSearchIngredientResults();
            }}
          >
            Add Ingredient
          </Button>
        </Link>

        <Link to="/search-ingredient">
          <Button
            bsSize="large"
            block
            style={{
              marginBottom: '20px'
            }}
          >
            Back
          </Button>
        </Link>
        {this.state.redirect && (
          <Redirect to="/add-ingredients" />
        )}
      </div>
    );
  }
  handleServingsChange(event) {
    const newServing = event.target.value;
    if (!isNaN(parseInt(newServing, 10))) {
      const newMacronutrients = this.state.originalDetails.macronutrients.map(macro => {
        return {
          ...macro,
          value: roundToTwo(parseFloat(macro.value) * parseInt(newServing, 10))
        };
      });
      const newMicronutrients = this.state.originalDetails.micronutrients.map(micro => {
        return {
          ...micro,
          value: roundToTwo(parseFloat(micro.value) * parseInt(newServing, 10))
        };
      });

      this.setState({
        servings: newServing,
        updatedDetails: {
          ...this.state.updatedDetails,
          macronutrients: newMacronutrients,
          micronutrients: newMicronutrients
        }
      });
    } else {
      this.setState({
        servings: newServing,
        updatedDetails: this.state.originalDetails
      });
    }
  }
  componentDidMount() {
    if (this.props.location.state.dbNumber) {
      searchNutrientInfo(this.props.location.state.dbNumber)
      .then(
        details => this.setState({
          originalDetails: details,
          updatedDetails: details
        })
      );
    } else if (this.props.location.state.ingredient) {
      this.setState({
        originalDetails: this.props.location.state.ingredient,
        updatedDetails: this.props.location.state.ingredient
      });
    }
  }
  render() {
    return (
      <div>
        {this.state.updatedDetails ? this.renderNutrientDetails() : 'Loading...'}
      </div>
    );
  }
};

IngredientDetails.propTypes = {
  addTempIngredient: PropTypes.func.isRequired,
  clearSearchIngredientResults: PropTypes.func.isRequired,
};

export default IngredientDetails;
