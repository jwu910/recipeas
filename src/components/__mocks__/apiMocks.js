export const nutrientInfo = {
  "name": "Fish, salmon, pink, raw",
  "measure": "3.0 oz",
  "macronutrients": [
    {
      "name": "Calories",
      "value": 108,
      "unit": "kcal"
    },
    {
      "name": "Protein",
      "value": 17,
      "unit": "g"
    },
    {
      "name": "Fat",
      "value": 3,
      "unit": "g"
    },
    {
      "name": "Carbohydrate",
      "value": 0,
      "unit": "g"
    }
  ],
  "micronutrients": [
    {
      "name": "Omega-3 Fatty Acids",
      "value": 0.48,
      "unit": "g"
    },
    {
      "name": "Calcium, Ca",
      "value": 6,
      "unit": "mg"
    },
    {
      "name": "Iron, Fe",
      "value": 0.32,
      "unit": "mg"
    },
    {
      "name": "Vitamin D (D2 + D3)",
      "value": 9.3,
      "unit": "µg"
    },
    {
      "name": "Vitamin B-12",
      "value": 3.53,
      "unit": "µg"
    },
    {
      "name": "Zinc, Zn",
      "value": 0.33,
      "unit": "mg"
    }
  ]
}
