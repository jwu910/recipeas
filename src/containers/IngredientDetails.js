import IngredientDetails from '../components/IngredientDetails';
import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';
import {
  addTempIngredient, clearSearchIngredientResults
} from '../actions/index';

const mapDispatchToProps = dispatch => {
  return bindActionCreators({
    addTempIngredient,
    clearSearchIngredientResults
  }, dispatch);
};

export default connect(null, mapDispatchToProps)(IngredientDetails);
