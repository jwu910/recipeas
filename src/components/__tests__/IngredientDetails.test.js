import React from 'react';
import IngredientDetails from '../IngredientDetails';
import renderer from 'react-test-renderer';
import { mount } from 'enzyme'; 
import { nutrientInfo } from '../__mocks__/apiMocks';
import { MemoryRouter } from 'react-router-dom';

describe('Edit Button', () => {
  const props = {
    addTempIngredient: jest.fn(),
    clearSearchIngredientResults: jest.fn(),
    location: {
      state: {
        dbNumber: 15083
      }
    },
    searchNutrientInfo: (foodKey) => new Promise((resolve, reject) => {
      return foodKey === 15083 ? resolve(nutrientInfo) :
      reject('Error!');
    }),
  };
  test('Snapshot test', () => {
    const wrapper = renderer.create(
      <MemoryRouter>
        <IngredientDetails {...props} />
      </MemoryRouter>
    );
    const tree = wrapper.toJSON();

    expect(tree).toMatchSnapshot();
  });

  test('Calls componentDidMount', () => {
    const spy = jest.spyOn(IngredientDetails.prototype, 'componentDidMount');
    mount(
      <MemoryRouter>
        <IngredientDetails {...props} />
      </MemoryRouter>
    );

    expect(spy).toHaveBeenCalled();
  });
});
