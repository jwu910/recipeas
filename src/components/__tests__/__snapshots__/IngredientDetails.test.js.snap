// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Edit Button Snapshot test 1`] = `
<div>
  Loading...
</div>
`;
